eclipse-collections-code-generator/src/main/resources/api/set/mutablePrimitiveSet.stg

```diff
@@ -88,6 +88,25 @@ public interface Mutable<name>Set extends Mutable<name>Collection, <name>Set
             return set.toSet().withAll(this);
         }
     }
+
+    /**
+     * Returns the set of all objects that are members of both {@code this} and {@code set}. The intersection of
+     * [1, 2, 3] and [2, 3, 4] is the set [2, 3].
+     *
+     * @since 11.0.
+     */
+    @Override
+    default Mutable<name>Set intersect(<name>Set set)
+    {
+        if (this.size() \< set.size())
+        {
+            return this.select(set::contains);
+        }
+        else
+        {
+            return set.asLazy().select(this::contains).toSet();
```

Comment:
why `asLazy()`? You should avoid `toSet()` by using the two argument `select` with a new set.

Reply:
Started out with the two argument approach like so - 

  `return set.select(this::contains, this.newEmpty());` 

Since the same won't work on the immutable side, I leaned towards symmetry. I will update the implementation on the mutable side to use select with a target empty collection and leave the lazy implementation on the immutable side. Let me know if that is good.

Reply:
why doesn't it work for the immutable?
``` set.select(this::contains, new <name>HashSet()).toImmutable()```

Reply:
It's in a default method, so can't reference the `<name>HashSet`. This would have to be implemented in the concrete class.

Reply:
It’s a major version change, we can add it on implementations. I would rather do that as select() and toSet() on Immutable sides is doubly expensive.

Reply:
Why is it doubly expensive? It's using lazy. It has an extra object creation for the primitive SelectIterable but we're creating a new primitive set anyway.

Reply:
The sequence `this.asLazy().select(that::contains).toSet().toImmutable()` on an immutable set first constructs some lazy stuff (cheap), then does the following:
- call `count` on `this` with the predicate being `that`. That's one loop through the whole thing.
- allocate an array of that size
- construct a `IfIntProcedure` object using `select` and `that::contains` as the predicate
- call `this.each(ifIntProc)`. This loops through the whole thing again.
- finally, the filled array is passed to the set constructor, where the result is allocated then filled by looping through the array.
- the contents of this is now copied to a second array.
- the second array is used to construct another set.
- this final set is wrapped as an immutable set.

so we compute the intersection twice and construct 3 temporary and 1 final place to put them (all the while recomputing hashes or copying stuff around).

It would take quite some effort to optimize that. Roughly, this would entail:
- stop using arrays -- use lists
- `toSet` and `toImmutable` should become lazy. this probably breaks everything, so the only realistic solution might be to have a `toImmutableSet`

PSA: don't use `asLazy` especially with immutables if memory, performance or sanity have any relevance to the code.

Reply:
Wow, this is a great bit of research and discovery @mohrezaei! Thank you. It looks like @nikhilnanivadekar you were correct this is doubly expensive. It's clear we should just override `intersect` in the concrete classes for immutable for now. 

We have a bunch of positive work we can do here as a result in separate PRs. I have been toying with the idea of adding `toImmutableSet` and friends. I might explore this to see some of the possibilities.

Reply:
Thank you @mohrezaei for the detailed explanation, will bear in my mind some of the points you highlighted with `asLazy` usages. 

@mohrezaei, @nikhilnanivadekar , @donraab    - 
Upon revisiting the solution, noticed that `toImmutable()` is available for me to use in combination with `select` on the immutable side Like so - 

        `return set.select(this::contains).toImmutable();`

Hoping this is closest to what we want to achieve in terms of implementation without compromising on optimization. Please let me know if this works! 

Thank you for reviewing the code so far.

Reply:
This looks good to me. The call `toImmutable() `will be a no-op if the set is already an `Immutable<name>Set`.

Reply:
@donraab It's not already immutable. I think we should really fix `toImmutable()` on primitive sets.
```java
        IntHashSet mutableSet = IntHashSet.newSetWith(this.toArray());
```
that line is the worst offender.  If we properly use the `copyOnWrite` flag in the set, we can even avoid all copying, which will make this code as efficient as possible:
```java
    @Override
    public ImmutableIntSet toImmutable()
    {
        if (this.size() == 0)
        {
            return IntSets.immutable.with();
        }
        if (this.size() == 1)
        {
            return IntSets.immutable.with(this.intIterator().next());
        }
        IntHashSet mutableSet;
        if (this.table.length == smallestPowerOfTwoGreaterThan(this.size() << 1)) // table is not too big
        {
            mutableSet = this;
            this.copyOnWrite = true;
        }
        else
        {
            IntHashSet mutableSet = new IntHashSet(this.size()) ;
            mutableSet.addAll(this);
        }
        return new ImmutableIntHashSet(mutableSet.table, mutableSet.occupiedWithData, mutableSet.zeroToThirtyOne, mutableSet.zeroToThirtyOneOccupied);
    }
```